### PR TITLE
Project initialization

### PR DESCRIPTION
**Pull request by mbarekh**

### CODE REVIEW

**Comment by mbarekh** [comment]
test this please

File: .travis.yml [hunk 1]
```diff
@@ -7,6 +7,8 @@ branches:
   - master
 os:
 - linux
+before_install:
+- yarn global install -g grunt-cli
```
Comment (Julien-Molina) [inline .travis.yml]: Could you put this in dev dependencies instead ?

Comment (mbarekh) [reply .travis.yml]: done

**Comment by mbarekh** [comment]
dist/ folder is present in most grafana projects. Probably, there is a reason for that. https://github.com/grafana/clock-panel

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+language: node_js
+node_js:
+- '8'
```
Comment (nicolaserny) [inline .travis.yml]: you can use a build matrix and test it on both node 8 and node 10

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+node_modules
```
Comment (nicolaserny) [inline .gitignore]: add yarn-debug.log
Update the file

You can get inspired from the github generated file ([Example](https://github.com/xcomponent/react-gojs/blob/master/.gitignore)

File: README.md [hunk 4]
```diff
@@ -1 +1,8 @@
-# PluginAC2
\ No newline at end of file
+[![Build Status](https://travis-ci.org/xcomponent/grafana-plugin-ac2.svg?branch=master)](https://travis-ci.org/xcomponent/grafana-plugin-ac2)
+
+## AC2 Panel Plugin for Grafana
+* AC2 Panel Plugin allows AC2 map visualization
```
Comment (nicolaserny) [inline README.md]: Add an installation section
[Example](https://github.com/grafana/simple-json-datasource)

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+sudo: false
```
Comment (nicolaserny) [inline .travis.yml]: Add a cache section
cache:
    directories:
        - node_modules

File: src/main.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
+import { PanelCtrl } from 'app/plugins/sdk';
+import moment from 'moment';
+import axios from './external/axios.min.js';
+import go from './external/go.js';
```
Comment (nicolaserny) [inline src/main.js]: We should use npm packages (go and axios)

**Comment by nicolaserny** [comment]
@Julien-Molina perhaps the dist folder should be updated by travis (deployment job)...

File: src/main.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
+import { PanelCtrl } from 'app/plugins/sdk';
```
Comment (nicolaserny) [inline src/main.js]: wrong file name:
should be AC2Ctrl.js instead of main.js

**Comment by nicolaserny** [comment]
> @mbarekh Are there any reasons not to use typescript ? ([Example](https://github.com/grafana/typescript-template-datasource)

All of our projects are in typescript. We should do the same with this plugin

**Comment by nicolaserny** [comment]
No tests ? Grafana should have guidelines about testing their plugins

**Comment by nicolaserny** [comment]
resolves #1

File: .travis.yml (hunk 2 above)
Comment (mbarekh) [reply .travis.yml]: done

File: .travis.yml (hunk 5 above)
Comment (mbarekh) [reply .travis.yml]: done

File: src/module.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  static templateUrl = "partials/template.html";
+
+  private panelDefaults = {
+    host: "localhost",
+    port: "7890",
+    application: "Servers,1.0.0",
+    user: "admin",
+    password: "admin"
+  };
+  private myFullDiagram: any = {
```
Comment (nicolaserny) [inline src/module.ts]: Do not use any!
Gojs has types such as Diagram...

File: src/module.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    div: undefined,
+    nodeTemplate: undefined,
+    groupTemplate: undefined,
+    linkTemplate: undefined,
+    model: undefined
+  };
+  private setupDiagramTimer: number | undefined = undefined;
```
Comment (nicolaserny) [inline src/module.ts]: Optional: private setupDiagramTimer?: number

File: src/module.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+  constructor($scope: any, $injector) {
+    super($scope, $injector);
+    this._panelConfig = new PanelConfig(this.panel);
+    _.defaultsDeep(this.panel, this.panelDefaults);
+
+    this.events.on('init-edit-mode', this._onInitEditMode.bind(this));
+  }
+
+  initDiagram() {
+    if ((<any>window).goSamples) (<any>window).goSamples();
```
Comment (nicolaserny) [inline src/module.ts]: What's goSamples ?

File: src/module.ts [hunk 11]
```diff
@@ -0,0 +1,157 @@
+import { PanelConfig } from './panel-config';
```
Comment (nicolaserny) [inline src/module.ts]: Split your class (too many features: creating diagram, rest calls, config etc...)

File: src/module.ts [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        'Content-Type': 'application/json',
+        'Accept': 'application/json'
+      },
+    })
+      .then(response => {
+        return axios.get(`${urlBase}/api/Application?application=${this.panel.application}&api_key=${response.data}`);
+      })
+      .then((response) => {
+        console.log(response.data);
+        const nodeDataArray: Array<any> = [];
```
Comment (nicolaserny) [inline src/module.ts]: Any...

File: src/module.ts [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        const groups = [];
+        response.data.forEach(component => {
+          if (groups.indexOf(component.GroupName as never) === -1) {
+            groups.push(component.GroupName as never);
+          }
+          const node: any = {};
```
Comment (nicolaserny) [inline src/module.ts]: any...

File: src/module.ts [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+          };
+          node.color = stateColor[component.State];
+          if (component.Parents.length > 0) {
+            node.parent = component.Parents[0];
+          }
+          nodeDataArray.push(node);
+        });
+        this.myFullDiagram.model = new go.TreeModel(nodeDataArray);
+      }).catch(error => {
+        console.log(error);
```
Comment (nicolaserny) [inline src/module.ts]: You should display an error message in the UI too

File: src/module.ts [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+            "InError": "gray",
+            "Starting": "Orange",
+          };
+          node.color = stateColor[component.State];
+          if (component.Parents.length > 0) {
+            node.parent = component.Parents[0];
+          }
+          nodeDataArray.push(node);
+        });
+        this.myFullDiagram.model = new go.TreeModel(nodeDataArray);
```
Comment (nicolaserny) [inline src/module.ts]: Question: as you create a new model every n seconds, does the UI blink and repaint all the nodes of the diagram ?

File: src/module.ts [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+      }).catch(error => {
+        console.log(error);
+      });
+  }
+
+  onClick() {
```
Comment (nicolaserny) [inline src/module.ts]: Can we have a more specific naming ? Click on which component ?

File: src/module.ts [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+  };
+  private setupDiagramTimer: number | undefined = undefined;
+  private _panelConfig: PanelConfig;
+
+  constructor($scope: any, $injector) {
+    super($scope, $injector);
+    this._panelConfig = new PanelConfig(this.panel);
+    _.defaultsDeep(this.panel, this.panelDefaults);
+
+    this.events.on('init-edit-mode', this._onInitEditMode.bind(this));
```
Comment (nicolaserny) [inline src/module.ts]: No dispose event ?

File: src/panel-config.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
+export class PanelConfig {
```
Comment (nicolaserny) [inline src/panel-config.ts]: Any...

File: src/partials/template.html [hunk 19]
```diff
@@ -0,0 +1,20 @@
+<head>
+    <title></title>
+    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
+    <style type="text/css">
```
Comment (nicolaserny) [inline src/partials/template.html]: Perhaps we can extract css definitions in a css file

File: src/partials/template.html [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  
+      #full {
+        min-height: 100%;
+      }
+  
+    </style>
+  </head>
+  
+  <body>
+    <div id="full">No map to display</div>
```
Comment (nicolaserny) [inline src/partials/template.html]: What's the meaning of the "full" id? Not an explicit name

File: tsconfig.json [hunk 21]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "strictNullChecks": true,
+    "module": "none",
+    "target": "es6",
+    "rootDir": "./src"
+  },
```
Comment (nicolaserny) [inline tsconfig.json]: Add exclude section

Exemple of tsconfig (to inspire)
```json
{
  "compilerOptions": {
    "outDir": "dist",
    "module": "commonjs",
    "target": "es5",
    "declaration": true,
    "lib": [
      "dom",
      "es6",
      "scripthost"
    ],
    "sourceMap": true,
    "allowJs": false,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "src",
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": false,
    "strictNullChecks": true,
    "suppressImplicitAnyIndexErrors": true,
    "noUnusedLocals": true,
    "skipLibCheck": true
  },
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "jest",
    "**/*.test.ts",
    "**/*.test.tsx"
  ],
  "include": [
    "src/**/*"
  ],
  "types": [
    "typePatches"
  ]
}
```

File: tsconfig.json [hunk 22]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "strictNullChecks": true,
+    "module": "none",
```
Comment (nicolaserny) [inline tsconfig.json]: Why none? I was expecting commonjs

File: tsconfig.json [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+    "target": "es6",
```
Comment (nicolaserny) [inline tsconfig.json]: Why es6? You should use es5 (to have better compatibility with old browsers

**Comment by nicolaserny** [comment]
@Julien-Molina grafana is based on angular. You should do a quick review too

File: package.json [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    "xcomponent",
+    "grafana",
+    "ac2"
+  ],
+  "author": "av",
+  "license": "Apache",
+  "repository": "https://github.com/xcomponent/grafana-plugin-ac2",
+  "devDependencies": {
+    "@types/angular": "^1.5.8",
+    "@types/grafana": "git+https://git@github.com/CorpGlory/types-grafana.git",
```
[the remark this answers is not in the document]
Comment (nicolaserny) [reply package.json]: Ok you updated the PR while I was reviewing.
Just ignore the comments when it's done 👍

File: tsconfig.json (hunk 23 above)
Comment (mbarekh) [reply tsconfig.json]: done

File: tsconfig.json (hunk 22 above)
Comment (mbarekh) [reply tsconfig.json]: done

File: tsconfig.json (hunk 21 above)
Comment (mbarekh) [reply tsconfig.json]: done

File: src/partials/template.html (hunk 20 above)
Comment (mbarekh) [reply src/partials/template.html]: It is the html div container for the map. I rename it to "container"

Comment (mbarekh) [reply src/partials/template.html]: done

File: src/partials/template.html (hunk 19 above)
Comment (mbarekh) [reply src/partials/template.html]: done

File: package.json [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+    "@types/lodash": "^4.14.74",
+    "babel-core": "^6.26.0",
+    "babel-loader": "^7.1.2",
+    "babel-preset-env": "^1.6.0",
+    "copy-webpack-plugin": "^4.0.1",
+    "loader-utils": "^1.1.0",
+    "ts-loader": "^2.3.7",
+    "typescript": "^2.4.2",
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply package.json]: done

File: package.json [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+  "keywords": [
+    "xcomponent",
+    "grafana",
+    "ac2"
+  ],
+  "author": "av",
+  "license": "Apache",
+  "repository": "https://github.com/xcomponent/grafana-plugin-ac2",
+  "devDependencies": {
+    "@types/angular": "^1.5.8",
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply package.json]: done

File: src/main.js (hunk 7 above)
Comment (mbarekh) [reply src/main.js]: done

File: README.md (hunk 4 above)
Comment (mbarekh) [reply README.md]: done

File: .gitignore (hunk 3 above)
Comment (mbarekh) [reply .gitignore]: done

File: src/module.ts [hunk 27]
```diff
@@ -0,0 +1,157 @@
+import { PanelConfig } from './panel-config';
+import { PanelCtrl, loadPluginCss } from 'grafana/app/plugins/sdk';
+import * as _ from 'lodash';
+import * as go from '../node_modules/gojs/release/go';
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/module.ts]: done

File: .travis.yml [hunk 28]
```diff
@@ -0,0 +1,17 @@
+language: node_js
+node_js:
+- '8'
+- '10'
+cache:
+directories:
```
Comment (nicolaserny) [inline .travis.yml]: I think the tabulation is wrong. (te be verified)
```suggestion
     directories:
          - node_modules
```

File: tsconfig.json (hunk 21 above)
Comment (nicolaserny) [reply tsconfig.json]: Missing important properties such as outDir etc...

File: .travis.yml (hunk 28 above)
Comment (mbarekh) [reply .travis.yml]: done

File: tsconfig.json (hunk 21 above)
Comment (mbarekh) [reply tsconfig.json]: done

File: src/module.ts (hunk 11 above)
Comment (mbarekh) [reply src/module.ts]: done. I created a ac2-map module to handle the gojs diagram part

File: src/module.ts [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+import * as axios from '../node_modules/axios/dist/axios';
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/module.ts]: done

File: src/module.ts (hunk 8 above)
Comment (mbarekh) [reply src/module.ts]: done

File: src/module.ts (hunk 9 above)
Comment (mbarekh) [reply src/module.ts]: done

File: src/module.ts (hunk 10 above)
Comment (mbarekh) [reply src/module.ts]: I removed that

File: src/module.ts (hunk 12 above)
Comment (mbarekh) [reply src/module.ts]: done

File: src/module.ts (hunk 13 above)
Comment (mbarekh) [reply src/module.ts]: done

File: src/module.ts (hunk 15 above)
Comment (mbarekh) [reply src/module.ts]: yes

File: src/module.ts (hunk 16 above)
Comment (mbarekh) [reply src/module.ts]: It is the button that loads the map. I renamed it "onClickLoadButton"

File: src/panel-config.ts (hunk 18 above)
Comment (mbarekh) [reply src/panel-config.ts]: done

File: src/module.ts [hunk 30]
```diff
@@ -0,0 +1,78 @@
...
+        return axios.get(`${urlBase}/api/Application?application=${this.panel.application}&api_key=${response.data}`);
+      });
+  }
+
+  setupDiagram() {
+    this.restCall()
+      .then(response => this.map.update(response))
+      .catch(error => {
+        console.error(error);
+        const container = document.getElementById("container");
```
Comment (nicolaserny) [inline src/module.ts]: The error handling is a bit low level (getElementIdById + innerHTML)
As grafana extensions are based on angular, it should be a better way to do that.
You can create a github issue to fix that in another PR

File: src/ac2-map.ts [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+            };
+            if (component.Parents.length > 0) {
+                linkDataArray.push({ from: component.Parents[0], to: node.key });
+            }
+            nodeDataArray.push(node);
+        });
+        groups.forEach(group => {
+            nodeDataArray.push({ key: group + "_group", text: group, isGroup: true });
+        });
+        this.diagram.model = new go.GraphLinksModel(nodeDataArray, linkDataArray);
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: done

File: src/ac2-map.ts [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+                        { font: "Bold 12px Sans-Serif", alignment: go.Spot.Center, margin: 4, stroke: "white" },
+                        new go.Binding("text", "text"))
+                ),
+                $(go.Placeholder,
+                    { padding: new go.Margin(0, 10) })
+            )
+        );
+    }
+
+    public update(response) {
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: done

File: src/ac2-map.ts [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+    public init() {
+        this.diagram = this.getDiagramTemplate();
+        this.diagram.nodeTemplate = this.getNodeTemplate();
+        this.diagram.groupTemplate = this.getGroupTemplate()
+        this.diagram.linkTemplate = this.getLinkTemplate();
+    }
+
+    private getDiagramTemplate() {
+        const $ = go.GraphObject.make;
+        return $(go.Diagram, "container",
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: done

File: src/ac2-map.ts [hunk 34]
```diff
@@ -0,0 +1,115 @@
+import * as go from 'gojs';
+
+export const stateColor = {
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: done

File: src/ac2-map.ts [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+
+    public init() {
+        this.diagram = this.getDiagramTemplate();
+        this.diagram.nodeTemplate = this.getNodeTemplate();
+        this.diagram.groupTemplate = this.getGroupTemplate()
+        this.diagram.linkTemplate = this.getLinkTemplate();
+    }
+
+    private getDiagramTemplate() {
+        const $ = go.GraphObject.make;
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: done

File: src/module.ts (hunk 30 above)
Comment (mbarekh) [reply src/module.ts]: done

File: src/module.ts (hunk 17 above)
Comment (mbarekh) [reply src/module.ts]: No dispose event is used in similar projects https://github.com/CorpGlory/grafana-progress-list/blob/master/src/module.ts

File: src/ac2-map.ts [hunk 36]
```diff
@@ -110,6 +122,28 @@ export class AC2Map {
...
+            linkDataArray
+        }
+    }
+
+    public draw(data: Array<AC2Data>) {
+        const goJsData = this.getGoJsData(data);
+        this.diagram.model = new go.GraphLinksModel(goJsData.nodeDataArray, goJsData.linkDataArray);
+    }
+
+    public update(data: Array<AC2Data>) {
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: done

File: src/ac2-map.ts [hunk 37]
```diff
@@ -20,20 +21,34 @@ export interface LinkDataArrayItem {
...
 
     private diagram: go.Diagram;
+    private containerId: string = "container";
 
     public init() {
-        this.diagram = this.getDiagramTemplate();
-        this.diagram.nodeTemplate = this.getNodeTemplate();
-        this.diagram.groupTemplate = this.getGroupTemplate()
-        this.diagram.linkTemplate = this.getLinkTemplate();
+        const $ = go.GraphObject.make;
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: done

File: src/ac2-map.ts [hunk 38]
```diff
@@ -20,20 +21,34 @@ export interface LinkDataArrayItem {
...
+        this.diagram.groupTemplate = this.getGroupTemplate($)
+        this.diagram.linkTemplate = this.getLinkTemplate($);
     }
 
-    private getDiagramTemplate() {
-        const $ = go.GraphObject.make;
-        return $(go.Diagram, "container",
+    public clear() {
+        if (this.diagram) {
+            this.diagram.div = null as any;
```
[the remark this answers is not in the document]
Comment (mbarekh) [reply src/ac2-map.ts]: I am getting some errors when using clear function instead

[the remark this answers is not in the document]
Comment (nicolaserny) [reply src/ac2-map.ts]: What kind of errors ?
Other idea: you can clear the model (set the empty arrays)

**Pull request merged**